### PR TITLE
Improve callbacks and adds tensorboard

### PR DESCRIPTION
**Pull request by zeyueN**
**Context:**
To facilitate the investigation of the optimization process on the numerics side, as well as giving users more control over the optimization progress, the callback functionality is expanded to allow for realtime monitoring and updates. A builtin callback that tracks optimization progresses in Tensorboard is also provided.

**Description of the Change:**

- The `callback` argument to `Optimizer.minimize` is now `callbacks` with the new possibility of passing multiple callbacks.
- Adds tagged parameter traversal `Parametrized.traverse_trainables()` which accumulates the path of tags while traversing. This is now used by the optimizer to get unique human interpretable identifiers to parameters (for better tracking).
- Callback functions now accept as argument the optimizer, the cost value, the cost function, and the trainable parameter values as well as their gradient. This gives users a lot of run time control, but it's not necessary to use them all.
- The execution of callbacks moved to after the backprop but before the parameter update (originally at the end of the step).
- A `Callback` base class for users to extend and inherit from.
- A `TensorboardCallbacks` that does automatic tracking of costs and parameters (values and gradients) in real time.

Misc changes:

- I added a `path` item to the testing workflow so that non-code changes won't trigger it.

**Benefits:**
- Better observability.
- Finer control.
- Opens up a lot of possibilities for optimization.

**Possible Drawbacks:**
Users could introduce too much overhead if the callbacks are too heavy, thus potentially slowing down the optimization.

**Related GitHub Issues:**
None

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -4,6 +4,13 @@ on:
     branches:
       - develop
   pull_request:
+    paths:
```
Comment (ziofil) [inline .github/workflows/tests.yml:7]: what is `paths` for?

File: mrmustard/training/callbacks.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
```
Comment (ziofil) [inline mrmustard/training/callbacks.py:1]: 2023?

File: mrmustard/training/callbacks.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    * `.update_grads`:
+        The custom grads modifyer, which is expected to return a list of parameter gradients
+        after modification, to be applied to the parameters.
+    * `.update_optimizer`:
+        The custom optimizer updater, which is expected to modify the optimizer inplace for
+        things like scheduling learning rates.
+
+    """
+
+    #: Custom tag for a callball instance to be used as keys in `Optimizer.callback_history`.
```
Comment (ziofil) [inline mrmustard/training/callbacks.py:103]:
```suggestion
    #: Custom tag for a callback instance to be used as keys in `Optimizer.callback_history`.
```

**Review by ziofil — APPROVED** [review]
Fantastic. This is indeed opening up lots of possibilities 🤩

File: .github/workflows/tests.yml (hunk 1 above)
Comment (zeyueN) [reply .github/workflows/tests.yml:7]: It's used to specify a list of files that should trigger the workflow when changes are made to them, so that a commit that changes say changelog only won't rerun tests.

File: mrmustard/training/callbacks.py (hunk 3 above)
Comment (zeyueN) [reply mrmustard/training/callbacks.py:103]: No idea how this happened 🤦
